x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java

```diff
@@ -1001,6 +999,67 @@ protected LogicalPlan rule(LogicalPlan plan) {
         }
     }
 
+
+    //
+    // Handle aggs in ORDER BY. To help folding any aggs not found in Aggregation
+    // will be pushed down to the Aggregate and then projected. This also simplifies the Verifier's job.
+    // Similar to Having however using a different matching pattern since HAVING is always Filter with Agg,
+    // while an OrderBy can have multiple intermediate nodes (Filter,Project, etc...)
+    //
+    private static class ResolveAggsInOrderBy extends AnalyzeRule<OrderBy> {
+
+        @Override
+        protected boolean skipResolved() {
+            return false;
+        }
+
+        @Override
+        protected LogicalPlan rule(OrderBy ob) {
+            List<Order> orders = ob.order();
+
+            // 1. collect aggs inside an order by
+            List<NamedExpression> aggs = new ArrayList<>();
+            for (Order order : orders) {
+                if (Functions.isAggregate(order.child())) {
+                    aggs.add(Expressions.wrapAsNamed(order.child()));
+                }
+            }
+            if (aggs.isEmpty()) {
+                return ob;
+            }
+
+            // 2. find first Aggregate child and update it
+
+            final AtomicBoolean found = new AtomicBoolean(false);
```

Comment:
Not it's an utility class for storing state outside a lamba expression. Since the class is already part of the JDK (so it's likely to be inlined) it's quite convenient. The alternative is to create our own Holder class.

Reply:
got it thx!

Reply:
I would opt for a `boolean[] found = new boolean[] {false}` instead, to avoid confusion and because AtomicBoolean has special semantics for the compiler to achieve the atomic operations in CPU level which is a bit too much, imho, just for convenience.